In [2]:
import numpy as np

## Matrix Basics

* This is how a m x n matrix A looks like: <br><br>
A = $
\begin{bmatrix}
\ a_{11} & a_{12} & \cdots & \ a_{1n} \\
\ a_{21} & a_{22} & \cdots & \ a_{2n} \\
\vdots & \vdots & \vdots &\vdots \\
\ a_{m1} & a_{m2} & \cdots & \ a_{mn} \\
\end{bmatrix}$

## Transpose Matrix

* If A is a m x n matrix, its transpose $A^{T}$ is a n x m mattrix <br><br>
A = $
\begin{bmatrix}
\ 2 & 3 \\
\ 1 & 4 \\
\ 5 & 6 \\
\end{bmatrix}$ &ensp; &ensp; &ensp; $A^{T}$ = $
\begin{bmatrix}
\ 2 & 1 & 5 \\
\ 3 & 4 & 6\\
\end{bmatrix}$

## Zero Matrix

* A matrix where all its elements are 0 <br><br>
A = $
\begin{bmatrix}
\ 0 & 0 \\
\ 0 & 0 \\
\ 50 & 0 \\
\end{bmatrix}$

## Identity Matrix
* A matrix where its diagonal is 1 and all the other elements are 0<br><br>

I = $
\begin{bmatrix}
\ 1 & 0 & 0 \\
\ 0 & 1 & 0 \\
\ 0 & 0 & 1 \\
\end{bmatrix}$

## Inverse Matrix
* If there exists matrix A and B such that AB = BA = I, A is an **invertible matrix** or **nonsingular matrix**, and B is A's **inverse matrix** <br><br>

A = $
\begin{bmatrix}
\ a & b \\
\ c & d \\
\end{bmatrix}$ &ensp; &ensp; &ensp; $A^{-1}$ = 
$\frac{1}{ad - bc}$
$
\begin{bmatrix}
\ d & -b \\
\ -c & a \\
\end{bmatrix}$<br><br>

* $ad - bc$ is called the determinant

## Matrix Transformation
* If A is a $m$ x $n$ matrix and $\vec{x}$ is a $R^n$ column vector, the product of A$\vec{x}$ is a $R^m$ column vector
<br><br>
* Example 1: Let A = $
\begin{bmatrix}
\ 1 & -1 \\
\ 2 & 5 \\
\ 3 & 4 \\
\end{bmatrix}$, &ensp; and $T_A$ is a transformation that maps $R^2$ 2x1 vector to a $R^3$ 3x1 vector
<br><br>
    * $A\vec{x}$ = $\begin{bmatrix}
\ 1 & -1 \\
\ 2 & 5 \\
\ 3 & 4 \\
\end{bmatrix}$ &ensp; $
\begin{bmatrix}
\ x_1 \\
\ x_2 \\
\end{bmatrix}$ = $\begin{bmatrix}
\ x_1 - x_2 \\
\ 2x_1 + 5x_2 \\
\ 3x_1 + 4x_2 \\
\end{bmatrix}$
<br><br>
    * $T_A$($x_1$, $x_2$) = ($x_1 - x_2$, $2x_1 + 5x_2$, $3x_1 + 4x_2$)
<br><br>
    * $T_A$($
\begin{bmatrix}
\ -1 \\
\ 3 \\
\end{bmatrix}$) = $
\begin{bmatrix}
\ -4 \\
\ 13 \\
\ 9 \\
\end{bmatrix}$

## Eigenvalue and Eigenvectors

* For a n by n matrix, if there exists a scala $\lambda$ and a non-zero vector $\vec{v}$ such at $A\vec{x}$ = $\lambda\vec{x}$, then $\lambda$ is $A$'s eigenvalue and $\vec{v}$ is the corresponding eigenvectors
* If $\lambda$ is $A$'s eigenvalue, then there are infinitely many eigenvectors

## Calculating Eigenvalue

* For a 2 by 2 matrix, det($A$ - $\lambda{I_2}$) = 0
* det($A$ - $\lambda{I_2}$) = $\lambda^2$ - $tr(A)\lambda$ + $det(A)$ = 0

## Similar Matrices

* For an invertible matrix $S$, if $S^{-1} A S$ = $B$, then square matrices A and B are called **similar**
* Similar matrices share same eigenvalues

## Matrix Diagonalization

* If a square matrix $A$ is similar to a diagonal matrix $D$, which means $S^{-1} A S$ = $D$, then $A$ is said to be **diagonalizable**
* This means that a n by n matrix $A$ has to have n linearly independent eigenvectors in order to be diagonalizable
* Diagonalizing a n by n matrix $A$ with n linearly independent eigenvectors:
    * Find $A$'s n linearly independent eigenvectors $\vec{s_1}$, $\vec{s_2}$, ..., $\vec{s_n}$
    * Form matrix $S$ = [$\vec{s_1}$,, $\vec{s_2}$, ..., $\vec{s_n}$]
    * The resulting matrix of $S^{-1} A S$ will be a diagonal matrix whose diagonal elements will be the corresponding eigenvalues

## Practice with Codes

Use Python's NumPy to find the followings:<br>
1. Handle numpy.linalg library 
2. Find eigenvectors

### 1. Finding Eigenvector

This is also refered to as **eigen decomposition**, which is the process of finding eigenvalues and eigenvectors

In [3]:
#create a 3x3 array
A = np.array([[1,2,3], [4,5,6], [7,8,9]])
A

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [4]:
np.linalg.eig(A)

(array([ 1.61168440e+01, -1.11684397e+00, -1.30367773e-15]),
 array([[-0.23197069, -0.78583024,  0.40824829],
        [-0.52532209, -0.08675134, -0.81649658],
        [-0.8186735 ,  0.61232756,  0.40824829]]))

In ndarray, the columns are eigenvectors, not the rows. For example column (-0.23197069, -0.52532209, -0.8186735) is one eigenvector

### 2. Eigenvectors Are Unit Vectors (Size is 1)

In [10]:
eigen_val, eigen_vec = np.linalg.eig(A)
display(eigen_val)
display(eigen_vec)

array([ 1.61168440e+01, -1.11684397e+00, -1.30367773e-15])

array([[-0.23197069, -0.78583024,  0.40824829],
       [-0.52532209, -0.08675134, -0.81649658],
       [-0.8186735 ,  0.61232756,  0.40824829]])

In [11]:
#each eigenvectors length is 1
np.sum(eigen_vec ** 2, axis=0)

array([1., 1., 1.])

### 3. Eigenvectors Are Orthogonal ONLY for Symmetric Matrix

In [12]:
#A is not a symmetric matrix
display(A)

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

Eigenvectors of A are therefore **not** orthogonal

In [14]:
display(np.dot(eigen_vec[:,0], eigen_vec[:,1]))
display(np.dot(eigen_vec[:,0], eigen_vec[:,2]))
display(np.dot(eigen_vec[:,1], eigen_vec[:,2]))

-0.27343437080986494

5.551115123125783e-17

-9.992007221626409e-16

In [15]:
#B is a symmetric maxtix
B = np.array([[1,-2], [-2,1]])
display(B)

array([[ 1, -2],
       [-2,  1]])

In [17]:
b_eigen_val, b_eigen_vec = np.linalg.eig(B)
display(b_eigen_val)
display(b_eigen_vec)

array([ 3., -1.])

array([[ 0.70710678,  0.70710678],
       [-0.70710678,  0.70710678]])

B is a symmetric matrix, therefore, its eigenvectors are orthogonal

In [20]:
np.dot(b_eigen_vec[:, 0], b_eigen_vec[:, 1])

0.0

### 4. Diagonalization

* If a square matrix $A$ and a diagonal matrix $D$ are similar matrices such that $S^{-1}AS = D$, then $A$ is said to be **diagonalizable**.
* Requirement is that a n by n matrix must have n linearly independent eigenvectors
* How to diagonalize a n by n matrix with n linearly independent eigenvectors:<br>
    1. Find $A$'s $n$ linearly independent eigenvectors $\vec{s_1}, \vec{s_2}, ..., \vec{s_n}$
    2. Matrix $S$ = [$\vec{s_1} \dots \vec{s_n}$]
    3. The product matrix of $S^{-1}AS$ will a diagonal matrix whose diagonal elements are corresponding eigenvalues of $\vec{s_1} \dots \vec{s_n}$


* Because this is so important, **remember**
    1. Diagonalization only exists for square matrices
    2. Eigenvectors must be linearly independent in order for the matrix to be diagonalized
    3. **Symmetric matrices** are diagonalizable



In [33]:
c = np.array([[1, 2, 3], [2, 4, 5], [3, 5, 6]])
c

array([[1, 2, 3],
       [2, 4, 5],
       [3, 5, 6]])

In [34]:
eigen_val, eigen_vec = np.linalg.eig(c)
display(eigen_val)
display(eigen_vec)

array([11.34481428, -0.51572947,  0.17091519])

array([[-0.32798528, -0.73697623,  0.59100905],
       [-0.59100905, -0.32798528, -0.73697623],
       [-0.73697623,  0.59100905,  0.32798528]])

In [35]:
np.diag(eigen_val)

array([[11.34481428,  0.        ,  0.        ],
       [ 0.        , -0.51572947,  0.        ],
       [ 0.        ,  0.        ,  0.17091519]])

In [36]:
np.linalg.inv(eigen_vec).dot(c.dot(eigen_vec))

array([[ 1.13448143e+01,  6.93889390e-16,  4.09394740e-16],
       [ 1.77635684e-15, -5.15729472e-01,  8.04911693e-16],
       [ 3.55271368e-15,  4.30211422e-16,  1.70915189e-01]])

In [37]:
eigen_vec.dot(np.diag(eigen_val)).dot(np.linalg.inv(eigen_vec))

array([[1., 2., 3.],
       [2., 4., 5.],
       [3., 5., 6.]])